File: Matlab Files/Network Generation/calc_Jacobian2.m
function [ J ] = calc_Jacobian2( nodes, fibers, init_lens, fibtype, fib_areas, rve_size  )
% calc_Jacobian calculates the numerical Jacobian of the nodal forces
%   w.r.t. perturbations to the nodal x,y,z coordinates. The Jacobian
%   is calculated as: J = d(F^i)/d(x^j) = d(mag(F)*n^i)/d(x^j) =
%   d(mag(F^i))/d(lambda^i) * d(lambda^i)/dx^j * n^i + d(n^i)/d(x^j) * mag(F^i).
%   F is the force between two fiber nodes, n^i is the normal vector
%   contribution in a given direction between two fiber, x^j is the
%   coordinate being probed, lambda is the fiber stretch between two nodes.
%   All derivative quantities are calculated numerically for dx prescribed
%   below.

% NOTE: The value of dx (aka epsilon) greatly affects the calculation of
%       the numerical Jacobian!

%   INPUTS: nodes - Nx3 matrix of nodal x,y,z coordinates
%           fibers - Mx5 vector of fiber connectivity
%           init_lens - Mx1 vector of fiber initial lengths
%           fibtype - Mx1 vector of fiber types
%           fib_areas - Mx1 vector of fiber cross-sectional areas
%           rve_size - 3x1 current rve size given the stretch of the
%           network

%   OUTPUTS: J - 3M x 3M Jacobian matrix dF^i/dx^j

% defining dx (epsilon)
dx = 1e-8 ;

J = zeros(3*size(nodes,1)) ;
nodes_n = nodes;

marg = 1e-12; % margin of error for bounds
up_bnd(1) = -0.5+rve_size(1,1)+marg ; % x-direction
up_bnd(2) = -0.5+rve_size(2,2)+marg ; % y-direction
up_bnd(3) = -0.5+rve_size(3,3)+marg ; % z-direction
lo_bnd = -0.5-marg ; % same bound for all directions

% tether force Jacobian (added for stability)
% NOTE: There is no dFi/dxj since we assume the force on one node is
% unaffected by the position of another.

% k = 1e-12 ; % proportionality constant for tether (arbitrary)
% for m = 1:length(nodes)
%     % dFi/dxi
%     J(3*m-2 , 3*m-2) = -2*k ;
%     J(3*m-1 , 3*m-1) = -2*k ;
%     J(3*m-0 , 3*m-0) = -2*k ;
%     
% end

for n = 1:length(fibers)
    
    node1 = fibers(n,1) ;
    node2 = fibers(n,2) ;
    
    % shifting node2 to real position
    for i = 1:3 %(1=x 2=y 3= z)
        current_node1 =  round(nodes(node1,i),12) ;
        current_node2 =  round(nodes(node2,i),12) ;
        k = 1; h = 1; p = 1 ; q = 1;
        while current_node1 > up_bnd(i) || current_node2 > up_bnd(i) || current_node1 < lo_bnd || current_node2 < lo_bnd
            if current_node1 > up_bnd(i)
                fibers(n,2+i) = fibers(n,2+i) - 1;
                nodes_n(node1,i) = nodes(node1,i) - k*rve_size(i,i) ;
                current_node1 = current_node1 - rve_size(i,i) ;
                k = k+1 ;
            elseif current_node1 < lo_bnd
                fibers(n,2+i)= fibers(n,2+i) + 1;
                nodes_n(node1,i) = nodes(node1,i) + p*rve_size(i,i) ;
                current_node1 = current_node1 + rve_size(i,i) ;
                p = p+1;
            end
            
            if current_node2 < lo_bnd
                fibers(n,2+i) = fibers(n,2+i) - 1;
                nodes_n(node2,i) = nodes(node2,i) + h*rve_size(i,i) ;
                current_node2 = current_node2 + rve_size(i,i) ;
                h = h+1;
            elseif current_node2 > up_bnd(i)
                fibers(n,2+i)= fibers(n,2+i) + 1;
                nodes_n(node2,i) = nodes(node2,i) - q*rve_size(i,i) ;
                current_node2 = current_node2 - rve_size(i,i) ;
                q = q+1;
            end
        end
    end
    
    realnode2 = nodes_n(node2,:) + fibers(n,3:5)*rve_size ; % location of node2 on real vector from node1 to node2
    
    % calculating original values
    vect = realnode2-nodes_n(node1,:) ;
    unit = vect./norm(vect) ; % fiber unit vector
    
    len = norm(vect);
    
    lambda = norm(vect)/init_lens(n) ;
    
    [ ~, node_1_force, node_2_force, dFdlam] = FiberConstEqn( vect, init_lens(n), fibtype(n), fib_areas(n) ) ;
           
    % node 1 and 2 force should be the same (opposite directions), we average them to mitigate
    % any slight differences
    force = (norm(node_1_force) + norm(node_2_force))/2;
    
     dlamdx1 = -1/(init_lens(n)*len).*vect;
     dlamdx2 = 1/(init_lens(n)*len).*vect;
     
     for i=1:3
         for j=1:3
             
             if i==j
                 dndx11 = -1/len + vect(i)^2/(len^3) ;
                 dndx12 = 1/len - vect(i)^2/(len^3) ;
                 dndx21 = -1/len + vect(i)^2/(len^3) ;
                 dndx22 = 1/len - vect(i)^2/(len^3) ;
             else
                 dndx11 = vect(i)*vect(j)/(len^3) ;
                 dndx12 = -vect(i)*vect(j)/(len^3) ;
                 dndx21 = vect(i)*vect(j)/(len^3) ;
                 dndx22 = -vect(i)*vect(j)/(len^3) ;
             end
                          
             % Calculate dFi/dxj
             j_temp11 = dFdlam * dlamdx1(j) * unit(i) + force * dndx11 ; 
             j_temp12 = dFdlam * dlamdx2(j) * unit(i) + force * dndx12 ; 
             j_temp21 = -dFdlam * dlamdx1(j) * unit(i) - force * dndx21 ; % unit vector (n) has opposite sign
             j_temp22 = -dFdlam * dlamdx2(j) * unit(i) - force * dndx22 ; % unit vector (n) has opposite sign
             
             % this is dFi/dxi
             J(3*node1-(3-i) , 3*node1-(3-j)) = J(3*node1-(3-i) , 3*node1-(3-j)) + j_temp11 ;
             % this is dFj/dxj
             J(3*node2-(3-i) , 3*node2-(3-j)) = J(3*node2-(3-i) , 3*node2-(3-j)) + j_temp22 ;
             
             % this is dFi/dxj
             J(3*node1-(3-i) , 3*node2-(3-j)) = J(3*node1-(3-i) , 3*node2-(3-j)) + j_temp12 ;
             % this is dFj/dxi
             J(3*node2-(3-i) , 3*node1-(3-j)) = J(3*node2-(3-i) , 3*node1-(3-j)) + j_temp21 ;
         end
     end
     
    
end

end

File: VisualStudioFiles/FEBIO_PLUGIN/calcJacobian2.cpp
/* calcJacobain.cpp calculates the nuumerical JAcobian for the periodic RVE

SUMMARY: This function calculates the analytical Jacobian for the RVE of fibers
defined by fiberConstEqn.cpp. It stores the Jacobian as a sparse matrix for 
analysis using the Sparse Solvers in Eigen.

CALLED FROM: networkSolver.cpp
CALLS ON: calcForcesPeriodic.cpp

INPUTS: nodes_n - Nx3 matrix of nodal coordinates
		nodes0 - Nx3 matrix of nodal coordinates for tethering
		fibers - Mx5 fiber matrix relating fiber end nodes to one another
		init_lens - Mx1 vector of fiber initial lengths
		fib_type - Mx1 vector of fiber types
		fib_areas - Mx1 vector of fiber cross-sectional areas
		rve_stretch - 3x1 vector of RVE prinicpal stretches
		fib_forces - Mx1 vector of fiber forces (initially empty)
		node_forces - Nx3 matrix of x,y,z components of forces on nodes (initially empty)
		num_fibers - M integer number of fibers
		num_nodes - N integer number of nodes
		bnd_nodes - Px3 coordinates of boundary nodes
		J - 3*N x 3*N Jacobian matrix of dF_i/dx_j where F_i is the directional nodal force and x_j is the directional coordinate 

OUTPUTS: None

NOTE: nodes_n, fibers, fib_forces, node_forces, bnd_nodes, and J are modified by this code.

Author: Ryan Mahutga - Barocas Research Group - University of Minnesota

Last Update: 04-05-2018

*/

#include <iostream>
#include <cmath>
#include "Eigen/Eigen"
#include "omp.h"
#include "calcForcesPeriodic.h"
#include "dataStructureDefinitions.h"

using namespace std;
using namespace Eigen;

void calcJacobian2(MatrixXd& nodes_n, MatrixXd& fibers_n, network& network_n, Matrix3d F, VectorXd& node_forces,
	int num_fibers, int num_nodes,  SparseMatrix<double>& J, VectorXi stab_node)
{
	VectorXd init_lens = network_n.init_lens; // fiber deformed initial lengths
	VectorXd fib_areas = network_n.fib_areas; // fiber deformed areas
	VectorXi fib_type = network_n.fib_type; // fiber types

	int xyz_num = 3 * num_nodes; // DOFs

	MatrixXd Jac = MatrixXd::Zero(xyz_num, xyz_num); //clear Dense Jacobian

	int node1, node2; // nodes numbers

	Vector3d realNode2, // nodal positions of node2
		old_vect, unit_old, // finer vector and unit vector
		node_force1, node_force2, //forces on node1 and node2 
		dlamdx1, dlamdx2; // dstretch/dcoordinate

		double fib_force, // fiber force
			dFdlam, lambda, // d forcce/ d stretch, stretch
			vect_len, // vector length
			dndx11, dndx12,  dndx21, dndx22, // d normal / dx 
			j_temp11,  j_temp12, j_temp21, j_temp22, // temporary jacobian entries
			inv_vect_len3; // 1/L^3

		for (int n = 0; n < num_fibers; n++)
		{
			node1 = (int)(fibers_n(n, 0) + 0.3); // first (static) node of fiber 
			node2 = (int)(fibers_n(n, 1) + 0.3); // second (dynamic) node of fiber (i.e. node that is moved according to fibers(m,2:4))

			// real nodal coordinates of nearest node along fiber from node1 to node2
			realNode2(0) = nodes_n(node2, 0) + fibers_n(n, 2)*F(0, 0) + fibers_n(n, 3)*F(0, 1) + fibers_n(n, 4)*F(0, 2);
			realNode2(1) = nodes_n(node2, 1) + fibers_n(n, 2)*F(1, 0) + fibers_n(n, 3)*F(1, 1) + fibers_n(n, 4)*F(1, 2);
			realNode2(2) = nodes_n(node2, 2) + fibers_n(n, 2)*F(2, 0) + fibers_n(n, 3)*F(2, 1) + fibers_n(n, 4)*F(2, 2);

			// vector from node1 to node2
			old_vect(0) = realNode2(0) - nodes_n(node1, 0); // fiber x span
			old_vect(1) = realNode2(1) - nodes_n(node1, 1); // fiber y span
			old_vect(2) = realNode2(2) - nodes_n(node1, 2); // fiber z span

			vect_len = old_vect.norm(); // fiber length

			inv_vect_len3 = 1 / (vect_len*vect_len*vect_len); // 1/L^3

			// fiber unit vector
			unit_old = old_vect/vect_len;

			// calling fiber constitutive equation to get fib_forces and dFdlam
			fiberConstEqn(old_vect, node_force1, node_force2, lambda, init_lens(n), 
				fib_type(n), fib_areas(n), fib_force, dFdlam);

			// dlam/dx values for node1 x
			dlamdx1 = -1.0 / (init_lens(n)*vect_len)*old_vect;

			// dlam/dx values for node2 x
			dlamdx2 = 1.0 / (init_lens(n)*vect_len)*old_vect;

			// residual node forces
			node_forces(3 * node1 + 0) = node_forces(3 * node1 + 0) + node_force1(0);
			node_forces(3 * node1 + 1) = node_forces(3 * node1 + 1) + node_force1(1);
			node_forces(3 * node1 + 2) = node_forces(3 * node1 + 2) + node_force1(2);

			node_forces(3 * node2 + 0) = node_forces(3 * node2 + 0) + node_force2(0);
			node_forces(3 * node2 + 1) = node_forces(3 * node2 + 1) + node_force2(1);
			node_forces(3 * node2 + 2) = node_forces(3 * node2 + 2) + node_force2(2);

			// calculating Jacobian values
			for (int i = 0; i < 3; i++)
			{
				for (int j = 0; j < 3; j++)
				{
					if (j == i)
					{
						dndx11 = -1.0 / vect_len + old_vect(i)*old_vect(i) *inv_vect_len3; // dn/dx for F node1 and x node1
						dndx12 = 1.0 / vect_len - old_vect(i)*old_vect(i) *inv_vect_len3; // dn/dx for F node1 and x node2
						dndx21 = dndx11; // dn/dx for F node2 and x node1
						dndx22 = dndx12; // dn/dx for F node2 and x node2
					}
					else
					{
						dndx11 = old_vect(i)*old_vect(j) * inv_vect_len3; // dn/dx for F node1 and x node1
						dndx12 = -1.0* old_vect(i)*old_vect(j) * inv_vect_len3; // dn/dx for F node1 and x node2
						dndx21 = dndx11; // dn/dx for F node2 and x node1
						dndx22 = dndx12; // dn/dx for F node2 and x node2
					}

					// Calculate dFi / dxj
					j_temp11 = dFdlam * dlamdx1(j) * unit_old(i) + fib_force * dndx11; // temporary jacoian value F node1 x node1
					j_temp12 = dFdlam * dlamdx2(j) * unit_old(i) + fib_force * dndx12; // temporary jacoian value F node1 x node2
					j_temp21 = j_temp12; //-dFdlam * dlamdx1(j) * unit_old(i) - fib_force * dndx21;  // temporary jacoian value F node2 x node1
					j_temp22 = -dFdlam * dlamdx2(j) * unit_old(i) - fib_force * dndx22;  // temporary jacoian value F node2 x node2

					// this is dFi / dxi
					Jac(3 * node1 + i, 3 * node1 + j) += j_temp11;
					// this is dFj / dxj
					Jac(3 * node2 + i, 3 * node2 + j) += j_temp22;
					// this is dFi / dxj
					Jac(3 * node1 + i, 3 * node2 + j) += j_temp12;
					// this is dFj / dxi
					Jac(3 * node2 + i, 3 * node1 + j) = Jac(3 * node1 + i, 3 * node2 + j);
				}
			}
		}

	// Constraining the stabilized node to their initial positions
	double max_J = 10*Jac.maxCoeff(); // maximum J
	for (int p = 0; p < 3; p++)
	{
		if (stab_node(p) > -1)
		{
			Jac(3 * stab_node(p) + 0, 3 * stab_node(p) + 0) = max_J;
			Jac(3 * stab_node(p) + 1, 3 * stab_node(p) + 1) = max_J;
			Jac(3 * stab_node(p) + 2, 3 * stab_node(p) + 2) = max_J;

			Jac(3 * stab_node(p) + 0, 3 * stab_node(p) + 1) = 0;
			Jac(3 * stab_node(p) + 0, 3 * stab_node(p) + 2) = 0;
			Jac(3 * stab_node(p) + 1, 3 * stab_node(p) + 0) = 0;
			Jac(3 * stab_node(p) + 1, 3 * stab_node(p) + 2) = 0;
			Jac(3 * stab_node(p) + 2, 3 * stab_node(p) + 0) = 0;
			Jac(3 * stab_node(p) + 2, 3 * stab_node(p) + 1) = 0;


			node_forces(3*stab_node(p) + 0) = 0;
			node_forces(3*stab_node(p)+ 1) = 0;
			node_forces(3*stab_node(p)+ 2) = 0;
		}
	}
	
	double drop_tol = 1e-18; // drop tolerance for sparse jacobian

	J = Jac.sparseView(drop_tol, 1); // setting to sparse matrix
}